plugins/xdebug/xdebug.cc

```diff
@@ -519,44 +494,36 @@ XScanRequestHeaders(TSCont /* contp */, TSEvent event, void *edata)
-
-        // dump on server request
-        auto send_req_dump = [](TSCont /* contp */, TSEvent event, void *edata) -> int {
-          TSHttpTxn txn = (TSHttpTxn)edata;
-          TSMBuffer buffer;
-          TSMLoc hdr;
-          if (TSHttpTxnServerReqGet(txn, &buffer, &hdr) == TS_SUCCESS) {
-            // re-add header "X-Debug: log-headers", but only once
-            TSMLoc dst = TSMimeHdrFieldFind(buffer, hdr, xDebugHeader.str, xDebugHeader.len);
-            if (dst == TS_NULL_MLOC) {
-              if (TSMimeHdrFieldCreateNamed(buffer, hdr, xDebugHeader.str, xDebugHeader.len, &dst) == TS_SUCCESS) {
-                TSReleaseAssert(TSMimeHdrFieldAppend(buffer, hdr, dst) == TS_SUCCESS);
-                TSReleaseAssert(TSMimeHdrFieldValueStringInsert(buffer, hdr, dst, 0 /* idx */, "log-headers",
-                                                                lengthof("log-headers")) == TS_SUCCESS);
-                log_headers(txn, buffer, hdr, "ServerRequest");
-              }
-            }
-          }
-          TSHttpTxnReenable(txn, TS_EVENT_HTTP_CONTINUE);
-          return TS_EVENT_NONE;
-        };
-        TSHttpTxnHookAdd(txn, TS_HTTP_SEND_REQUEST_HDR_HOOK, TSContCreate(send_req_dump, nullptr));
-
-        // dump on server response
-        auto read_resp_dump = [](TSCont /* contp */, TSEvent event, void *edata) -> int {
-          TSHttpTxn txn = (TSHttpTxn)edata;
-          TSMBuffer buffer;
-          TSMLoc hdr;
-          if (TSHttpTxnServerRespGet(txn, &buffer, &hdr) == TS_SUCCESS) {
-            log_headers(txn, buffer, hdr, "ServerResponse");
-          }
-          TSHttpTxnReenable(txn, TS_EVENT_HTTP_CONTINUE);
+
+      } else if (header_field_eq("probe", value, vsize)) {
+        xheaders |= XHEADER_X_PROBE_HEADERS;
+
+        // prefix request headers and postfix response headers
+        BodyBuilder *data = new BodyBuilder();
+        data->txn         = txn;
+
+        TSVConn connp = TSTransformCreate(body_transform, txn);
+        TSContDataSet(connp, data);
+        TSHttpTxnHookAdd(txn, TS_HTTP_RESPONSE_TRANSFORM_HOOK, connp);
+
+        // store data pointer in txnarg to use in global cont XInjectResponseHeaders
+        TSHttpTxnArgSet(txn, BodyBuilderArgIndex, data);
+
+        // create a self-cleanup on close
+        auto cleanupBodyBuilder = [](TSCont /* contp */, TSEvent event, void *edata) -> int {
+          TSHttpTxn txn     = (TSHttpTxn)edata;
+          BodyBuilder *data = static_cast<BodyBuilder *>(TSHttpTxnArgGet(txn, BodyBuilderArgIndex));
+          delete data;
           return TS_EVENT_NONE;
         };
-        TSHttpTxnHookAdd(txn, TS_HTTP_READ_RESPONSE_HDR_HOOK, TSContCreate(read_resp_dump, nullptr));
+        TSHttpTxnHookAdd(txn, TS_HTTP_TXN_CLOSE_HOOK, TSContCreate(cleanupBodyBuilder, nullptr));
+
+        // disable writing to cache because we are injecting data into the body.
+        TSHttpTxnReqCacheableSet(txn, 0);
+        TSHttpTxnRespCacheableSet(txn, 0);
```

Comment:
I know it’s a mess , but do we need all of these set?

Reply:
Maybe. I remember I didn't want to cache the injected trace data, ever, because the upsteam might have trace data injected too. And I needed the transform to run, but not cache.